File: 2018/code/nim/day20.nim
import strutils

const d = readFile("../../input/input_day20.txt")

type
    FrameTile = char
    FrameValues = Frame
    Frame = seq[FrameSequence]
    FrameSequence = seq[FrameTile|FrameValues]
    
var frame: Frame
echo $frame

# proc generate(regex: string): Frame =
#     var frames = @[result]
#     for c in regex:
#         if "^$".contains(c):
#             continue
#         elif "NESW".contains(c):
#             let n = FrameObject(kind: FrameTile, tile: c)
#             frames[^1][^1].add(n)
#         elif c == '(':
#             let n = FrameObject(kind: FrameValues)
#             frames[^1][^1].add(n)
#             frames.add(n)
#         elif c == '|':
#             let n = FrameObject(kind: FrameValues)
#             frames[^1].add(n)
#         elif c == ')':
#             discard frames.pop()

# echo generate("^ENNWSWW(NEWS|)SSSEEN(WNSE|)EE(SWEN|)NNN$")

File: 2018/code/haskell/day20.hs
type Frame = [Segment]

type Segment = [Section]

data Section = F Frame | C Char deriving Show

processInput :: String -> Frame
processInput s = 
    let frame = [[]]
    in processFrame s frame [frame]

processFrame :: String -> Frame -> [Frame] -> Frame
processFrame (h:t) f fl
    | null t = nf
    | otherwise = processFrame' t 
    where
        (nf, nfl) = processFrame' h f fl

processFrame' :: Char -> Frame -> [Frame] -> (Frame, [Frame])
processFrame' c f@(hf, tf) fl@(hfl, tfl)
    | c == '|' = (f, tfl)
    | c == '(' = 
    | c == ')' =
    | any (\l => l == c) "NESW" = 
    | otherwise = 
    where


main :: IO()
main = do
    file <- readFile "../../input/input_day20.txt"
    let frames = processInput file